timemachine/cpp/src/kernels/k_exchange.cuh

```diff
@@ -11,6 +14,33 @@ template <typename RealType> RealType __host__ __device__ convert_nan_to_inf(con
     return input;
 }
 
+void __global__ k_setup_sample_atoms(
+    const int num_samples,
+    const int sample_atoms,          // number of atoms in each sample
+    const int *__restrict__ samples, // [num_samples]
+    const int *__restrict__ target_atoms,
+    const int *__restrict__ mol_offsets,
+    int *__restrict__ output_atom_idxs,
+    int *__restrict__ output_mol_offsets) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (idx < num_samples) {
+        int mol_idx = samples[idx];
+        int mol_start = mol_offsets[mol_idx];
+        int mol_end = mol_offsets[mol_idx + 1];
+        output_mol_offsets[mol_idx] = target_atoms[mol_start];
+        output_mol_offsets[mol_idx + 1] = target_atoms[mol_end - 1] + 1;
+        int num_atoms = mol_end - mol_start;
+
+        assert(num_atoms == sample_atoms);
```

Comment:
are `asserts` a no-op when compiling with `-O3`?

Reply:
They are not, this does trigger with `-O3`